osgp/protocol-adapter-oslp/web-device-simulator/src/main/java/org/opensmartgridplatform/webdevicesimulator/service/OslpChannelHandler.java

```diff
@@ -871,13 +919,25 @@ private void handleUpdateFirmwareRequest(final Device device, final UpdateFirmwa
         LOGGER.debug("handle UpdateFirmwareRequest for device: {}, with serialized size of {}",
                 device.getDeviceIdentification(), request.getSerializedSize());
 
-        this.sendDelayedDeviceRegistration(device);
-
-        // Send a firmware activation event after the device registration has
-        // (likely) been finished
-        final Oslp.Event event = Oslp.Event.FIRMWARE_EVENTS_ACTIVATING;
-        final String description = "A new firmware is activated";
-        this.sendEventWithCustomDelay(device, event, description, 7000);
+        this.simulateReboot(device).thenAccept(deviceMessageStatus -> {
+            if (DeviceMessageStatus.NOT_FOUND == deviceMessageStatus) {
+                LOGGER.error("Device {} not found handling update firmware request", device.getDeviceIdentification());
+                return;
+            }
+            final Oslp.Event event;
+            final String description;
+            if (DeviceMessageStatus.OK == deviceMessageStatus) {
+                event = Oslp.Event.FIRMWARE_EVENTS_ACTIVATING;
+                description = "A new firmware is activated";
+            } else {
+                LOGGER.warn("Device {} has issues registering after reboot handling update firmware request",
+                        device.getDeviceIdentification());
+                event = Oslp.Event.FIRMWARE_EVENTS_DOWNLOAD_FAILED;
+                description = "Failure rebooting and registrating device during firmware activation";
+            }
+            this.sendEventWithCustomDelay(device, event, description,
+                    Math.max(1, 7 - this.deviceManagementService.getRebootDelay()));
```

Comment:
I see what you are doing here, but on the other hand, this would  be an oppurtunity to get rid of the hard coded 7 second delay for the FIRMWARE_EVENTS_ACTIVATING event. I agree with you, that it's out of scope for this story, but maybe getting the 7 from a property file is not much work. I leave it up to you what to do with it. For me it's fine to leave it as is right now.

Reply:
In case the 7 sec should be configurable, it should be another user story.